tests/extension/oneapi_device_global/device_global_exceptions.cpp

```diff
@@ -0,0 +1,293 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Provides tests for the exceptions that are thrown by sycl::queue and
+//  sycl::handler copy member functions overloads
+//
+*******************************************************************************/
+
+#include "../../common/common.h"
+#include "../../common/get_cts_object.h"
+#include "device_global_common.h"
+#include "type_pack.h"
+
+#define TEST_NAME device_global_exceptions
+
+namespace TEST_NAMESPACE {
+using namespace sycl_cts;
+using namespace sycl_cts::util;
+using namespace device_global_common_functions;
+
+#if defined(SYCL_EXT_ONEAPI_PROPERTY_LIST) && \
+    defined(SYCL_EXT_ONEAPI_DEVICE_GLOBAL)
+namespace oneapi = sycl::ext::oneapi;
+
+/**
+ * @brief The functions checks if error code of the given exception is equal to
+ * sycl::errc::invalid
+ * @param e The sycl::exception instance that needs to verify error code
+ * @return True if error code of exception is sycl::errc::invalid, false
+ * otherwise
+ */
+inline bool is_errc_invalid(const sycl::exception& e) {
+  return e.code() != sycl::errc::invalid;
+}
+
+namespace handler_copy_exceptions {
+// Creating instance with default constructor
+template <typename T>
+oneapi::device_global<T> dev_global;
+
+/** @brief The function tests that .copy() member function overload throws an
+ * exception with error code equal to errc::invalid if attempt to write beyond
+ * the end of the destination variable
+ *  @tparam T Type of underlying device_global value
+ */
+template <typename T, size_t N>
+void run_test(util::logger& log, const std::string& type_name) {
+  std::shared_ptr<T[N]> src(new T[N]);
+  std::shared_ptr<T[N]> dest(new T[N]);
+
+  bool is_exception_thrown{true};
+  bool is_exception_correct{true};
+  auto queue = util::get_cts_object::queue();
+
+  queue.submit([&](sycl::handler& cgh) {
+    // Try to write data beyond the end of the destination variable
+    // Have to throw an exception
+    try {
+      cgh.copy<T>(src.get(), dev_global<T>, N, N / 2);
+      is_exception_thrown = false;
+    } catch (sycl::exception const& e) {
+      is_exception_correct = is_errc_invalid(e);
+    }
+
+    try {
+      cgh.copy<T>(dev_global<T>, dest.get(), N, N / 2);
+      is_exception_thrown = false;
+    } catch (sycl::exception const& e) {
+      is_exception_correct &= is_errc_invalid(e);
+    }
```

Comment:
Please, move copy operations to different command group scopes.

From https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:command.group.scope

> A [command group scope](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#command-group-scope), as defined in [Section 3.7.1](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:executionmodel), may execute a single [command](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#command) such as invoking a kernel, copying memory, or executing a host task. It is legal for a [command group scope](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#command-group-scope) to statically contain more than one call to a [command](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#command) function, but **_any single execution of the [command group function object](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#command-group-function-object) may execute no more than one [command](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#command)_**.

Reply:
Do we need to apply this change to [this place](https://github.com/KhronosGroup/SYCL-CTS/blob/f8d877560d82e2e4a7eacefe9d252178a7d602c1/tests/extension/oneapi_device_global/device_global_handler_copy_memcpy_overloads.cpp#L72-L85) too?

Reply:
Yes. The spec rule applies to any SYCL code.

Reply:
Okay. I'll make PR for it.

Reply:
cf https://github.com/KhronosGroup/SYCL-Docs/issues/223